cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -124,22 +133,90 @@ public PartitionedFileSetDataset(DatasetContext datasetContext, String name,
   @Override
   public void startTx(Transaction tx) {
     partitionsAddedInSameTx.clear();
+    operationsInThisTx.clear();
     super.startTx(tx);
     this.tx = tx;
   }
 
   @Override
   public void postTxCommit() {
+    // simply delete the quarantine directory for this transaction
+    try {
+      Location quarantine = getQuarantineLocation();
+      if (quarantine.exists()) {
+        boolean deleteSuccess = quarantine.delete(true);
+        if (!deleteSuccess) {
+          throw new DataSetException(String.format("Error deleting quarantine location %s.", quarantine));
+        }
+      }
+    } catch (IOException e) {
+      throw new DataSetException(String.format("Error deleting quarantine location for tx %s.", tx.getWritePointer()),
+                                 e);
+    }
+
     this.tx = null;
     super.postTxCommit();
   }
 
   @Override
   public boolean rollbackTx() throws Exception {
+    // rollback all the partition add and drop operations, in reverse order
+    // if any throw exception, suppress it temporarily while attempting to roll back the remainder operations
+    IOException ioe = null;
+    for (int i = operationsInThisTx.size() - 1; i >= 0; i--) {
+      PathOperation operation = operationsInThisTx.get(i);
+      try {
+        rollbackOperation(operation);
+      } catch (IOException e) {
+        if (ioe == null) {
+          ioe = e;
+        } else {
+          ioe.addSuppressed(e);
+        }
+      }
+    }
+    if (ioe != null) {
+      throw ioe;
+    }
+
     this.tx = null;
     return super.rollbackTx();
   }
 
+  private void rollbackOperation(PathOperation operation) throws IOException {
+    switch (operation.getOperationType()) {
+      case CREATE:
```

Comment:
Actually, there's an existing JIRA for this. I think that can come separately, since it should go through testing and will take more time because of that.
I pulled the existing JIRA out of the backlog, into 3.5: https://issues.cask.co/browse/CDAP-1393